File: Homework 2/Q2.py
##Generation of exponential random variable from uniform trials using inverse cdf method
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats

N=1000
u=np.zeros(N)
x=np.zeros(N)
theta=0.2   #Value of average waiting time

#Generation of exponential random variable from uniform trials using inverse cdf method
uniform_trials=np.array(np.random.uniform(size=1000))
expo_observed_values=-theta * np.log(1-uniform_trials)

#Generation of exponential randon variable using in-built function
expo_expected_values=np.array(np.random.exponential(theta,size=1000))

#Taking the x axis values of expected and observed random variables
expected=plt.hist(expo_expected_values,bins=30)[0]
observed=plt.hist(expo_observed_values,bins=30)[0]

#Running the chi-square goodness of fit test
[chi_square,p]=scipy.stats.chisquare(observed[0:15], expected[0:15])
print('Value of chi-square')
print (chi_square)
print('Value of p')
print (p)
#Running the ks goodness of fit test
[KS,p]=scipy.stats.kstest(expo_observed_values, 'expon')
print('Value of KS')
print (KS)
print('Value of p')
print (p)

#plt.figure()
#plt.hist(expo_expected_values)

#To calculate the number of exponential time intervals in 1 unit time
sum=0
count=0
count_array=[]
for i in range(0,1000):
    sum=sum+expo_observed_values[i]
    count=count+1
    if sum>1:
        count_array.append(count)
        count=0
        sum = 0
        
#Plotting the histogram of the number of exponential time intervals in 1 unit time
plt.figure()
plt.suptitle('Number of exponentially distributed time intervals in 1 unit time')
plt.xlabel('Number of exponential time intervals in 1 unit time')
plt.ylabel('Count of exponential time intervals in 1 unit time')
plt.hist(count_array)

File: Homework1/Q2.m
clc;

%Initialization of variables
k = 50;    %k = 5,10,30,50
prob = 0.5;
samp = 300;
count1 = zeros(samp);
count_total = zeros(k+1,1);

%Generate random numbers and count number of successes in all the trials
for j = 1:samp
    count = 0;
    for i = 1:k
        A(i) = rand(1);
        if A(i) > prob
        count = count + 1;
        end
    end
    count1(j) = count; 
end

%Count of Successes in all the samples
for j = 1:k+1
    for i = 1:samp
       if count1(i) == j-1
           count_total(j,1) = count_total(j,1)+1;
       end
    end
end

%Plotting histogram
figure(1);
bar(0:k,count_total);
xlim([0,k]);
xlabel('Value of k');
ylabel('Count');
title('Counting Successes for Bernoulli Random Variable');


